### PR TITLE
add node shutdown taint

### LINKED ISSUE #58635: Propose to taint node "shutdown" condition
Today there is only one node condition to represent whether kubelet is healthy (updating node status). Node controller will mark node as "NotReady" after kubelet stops updating for a while. The reason for kubelet not responding could be 
1. Network disconnect
2. Node is shun down (stopped) or paused
3. Kubelet is killed (could not response)

## Problem
In many cases, when a node becomes "NotReady", pods will be evicted, and new pods will be started on other nodes (failover for high availability). However, if node is shut down, the volumes attached to the nodes are still attached. Volume controller should try to detach volumes when pods are killed from the old node. But there is a safety check before detaching to see whether volume is still mounted. This check is done by a VolumeInUse field in node status which is updated by kubelet. In case of node shut down, this field will not be updated by kubelet so that the safety check will fail and assume the volume cannot be detached because of existing mounts. The new pods will be pending for a while because volume cannot be attached to the other node. Although the reality is that volume can be detached in this situation. (Note: some cloud provider will treat stopped node as not exist in cloud provider so that node controller will delete the node API object. But still detach will not happen because of the mount safety check. Ref https://github.com/kubernetes/kubernetes/issues/46442)

## Proposal
We propose to taint the node with "Stopped" condition. When kubelet stops updating, node controller checks cloud provider and see whether the node is in "stopped" state (Note: different cloud provider might use different term to represent this state, e.g., terminated). When volume controller gets a node update/delete event, it can check the node taint spec and mark the volumes attached to that node are safe to detach (volumes can be be removed from the list of VolumeInUse (meaning the volumes are mounted). When detach is issued, the operation can be proceeded without delay.

To determine the "stopped" condition, I lists a few cloud provider's explanation of stopping an instance.

### GCE

- Stopping an instance causes Compute Engine to send the ACPI Power Off signal to the instance. Modern guest operating systems are configured to perform a clean shutdown before powering off in response to the power off signal. Compute Engine waits a short time for the guest to finish shutting down and then transitions the instance to the TERMINATED state.

- You can stop an instance temporarily so you can come back to it at a later time. A stopped instance does not incur charges, but all of the resources that are attached to the instance will still be charged. Alternatively, if you are done using an instance, delete the instance and its resources to stop incurring charges.

- You can permanently delete an instance to remove the instance and the associated resources from your project. If the instance is part of an instance group, the group might try to recreate the instance to maintain a certain group size.

### AWS

- You can only stop an Amazon EBS-backed instance. The instance performs a normal shutdown and stops running; its status changes to stopping and then stopped. Any Amazon EBS volumes remain attached to the instance, and their data persists. In most cases, the instance is migrated to a new underlying host computer when it's started. 

- You can delete your instance when you no longer need it. This is referred to as terminating your instance. When an instance terminates, the data on any instance store volumes associated with that instance is deleted. If your instance is in an Auto Scaling group, the Auto Scaling service marks the stopped instance as unhealthy, and may terminate it and launch a replacement instance.

### OpenStack

- Admins can pause and unpause a Nova compute instance. When an instance is paused, the entire state of the instance is kept in RAM. Pausing an instance will disable access to that instance, but won't free up any of its resources. Another option is to suspend, and then resume, an instance. Like paused OpenStack instances, a suspended instance keeps its current state, but it is written to storage. 

- A third option is to shelve OpenStack instances.A shelved instance is actually shut down, which is not the case for suspended or paused instances. If admins decide they no longer need a shelved instance, they can remove it, which ensures that it doesn't maintain any hypervisor-level resources in use.

- The last option is to stop an instance in Nova, which will disconnect all of its associated resources. This means admins can't restore a stopped instance to its previous state. This option is only useful for OpenStack instances that an organization no longer needs. In all other cases, admins should shelve, suspend or pause the instance.

So basically a "stopped" node means that it is shut down like an operating system shut down situation. All mounts are gone when machine is shut down.

cc @yujuhong @dchen1107 @Random-Liu

### PR DESCRIPTION
**Pull request by zetaab**
**What this PR does / why we need it**: we need node stopped taint in order to detach volumes immediately without waiting timeout. More info in issue ticket #58635 

**Which issue(s) this PR fixes** 
Fixes #58635

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by zetaab** [comment]
Questions/comments:

- can we modify Instances interface?
- I need help implementing at least gce/aws part of InstanceShutdownedByProviderID
- the main point why we need this is because of volume detach things. Should that be separate PR or also in this one? After this node label thing is added we should somehow query node taints https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L202 here. If node has shutdown taint, we can skip waiting time. How we should query that? Query node state from apiserver using kubeclient? @jingxu97 
- this needs cloud-controller-manager in use. However, the current situation is that products like kops / openshift are not using cloud-controller-manager. If cloud-controller-manager is not in use, the taints are not added.

File: pkg/controller/cloud/node_controller.go
Comment (yastij) [inline pkg/controller/cloud/node_controller.go]: Adding the taints means that node is shutdown, no need for a value

File: pkg/controller/cloud/node_controller.go
Comment (yastij) [inline pkg/controller/cloud/node_controller.go]: Is this an update or a patch ?

File: pkg/scheduler/algorithm/well_known_labels.go
Comment (yastij) [inline pkg/scheduler/algorithm/well_known_labels.go]: We should choose between shutdown or stopped to be consistent

**Review by yastij — COMMENTED** [review]
I did a quick pass, I'll do another one tomorrow

File: pkg/scheduler/algorithm/well_known_labels.go
Comment (zetaab) [reply pkg/scheduler/algorithm/well_known_labels.go]: maybe shutdown is better word for this

File: pkg/controller/cloud/node_controller.go
Comment (zetaab) [reply pkg/controller/cloud/node_controller.go]: patch would be better, but then I must implement more code. For instance https://github.com/zetaab/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L366 is using update as well in this file.

Comment (yastij) [reply pkg/controller/cloud/node_controller.go]: There's #56350 to fix this, also there was some constrains that made us use update. If you use update you might overwrite Status, which is something we don't want.

**Comment by zetaab** [comment]
@yastij check last commit, it is now modified to use patch instead of update.

**Comment by jhorwit2** [comment]
@zetaab Lets please keep the implementation of clouds out of this PR and focus on the API change & node controller changes.

**Comment by zetaab** [comment]
@jhorwit2 so you want that I remove that openstack implementation as well? Is there now some code which needs to be removed? Could you comment more specifically. Now this solution is working one in openstack.

**Comment by jhorwit2** [comment]
@zetaab Yeah, for this PR lets focus on the implementation / api changes and save any cloud specific changes for future PR's since not all clouds treat stopped the same (https://github.com/kubernetes/kubernetes/issues/46442). We'll want to track that for each provider and work with the respective cloud providers to update the functionality so it aligns. We should return the [unimplemented error](https://github.com/zetaab/kubernetes/blob/58c41f60560f47cd68a09ebe317d573e0b82dab4/pkg/cloudprovider/cloud.go#L186) and handle it gracefully in the node controller to provide backwards compatibility.

**Comment by zetaab** [comment]
@jhorwit2 when you are talking about node controller do you mean this new cloud controller? Because I am not sure should I implement something to old kube-controller? Currently if people do not use cloud controller node monitoring is handled by kube-controller. Should it work there as well or is this cloud controller enough?

File: pkg/cloudprovider/providers/openstack/openstack_instances.go
Comment (yastij) [inline pkg/cloudprovider/providers/openstack/openstack_instances.go]: This change shouldn't be included in this PR as @jhorwit2 said

**Comment by jhorwit2** [comment]
> when you are talking about node controller do you mean this new cloud controller?

@zetaab I'm referring to the cloud node controller you made the changes to in this PR already (`pkg/controller/cloud/node_controller.go`)

File: pkg/controller/cloud/node_controller.go
Comment (jhorwit2) [inline pkg/controller/cloud/node_controller.go]: You should use [`AddOrUpdateTaintOnNode`](https://github.com/kubernetes/kubernetes/blob/d6325933e1b5ac8a2c0a5a65ab8bd690015f4697/pkg/controller/controller_utils.go#L893) and remove most of this logic including the patch method and undo the changes to `getCloudTaint`.

**Review by jhorwit2 — CHANGES_REQUESTED** [review]

**Comment by jsafrane** [comment]
cc: @jingxu97 @gnufied

**Comment by jdumars** [comment]
cc: @khenidak

File: pkg/controller/cloud/node_controller.go
Comment (yastij) [inline pkg/controller/cloud/node_controller.go]: can we have some tests for this ?

File: pkg/controller/cloud/node_controller.go
Comment (gmarek) [inline pkg/controller/cloud/node_controller.go]: This means that the semantics or this Taint are somewhat cloudy. 
- It's applied when Node is NotReady and VM is in a given state
- It's removed when Node is Ready

It may be correct, but I guess full negation would be better, i.e.
- It's removed when Node is Ready or VM is not in the given state

Comment (zetaab) [reply pkg/controller/cloud/node_controller.go]: You are correct that it would be better to update that all the time. However, if we do that we will make request to cloudprovider api every 5 seconds. 99% of these requests are not needed. State is not changed. If we have 100 node clustee it maked 100 api requests every 5 second. Imo too mich because its not really needed

File: pkg/controller/cloud/node_controller.go
Comment (zetaab) [reply pkg/controller/cloud/node_controller.go]: I can add those when i have time. Hopefully today/tomorrow

File: pkg/controller/cloud/node_controller.go
Comment (gmarek) [reply pkg/controller/cloud/node_controller.go]: Instead of asking about every single VM we can just ask for the list of all VMs and work with that.

Comment (jhorwit2) [reply pkg/controller/cloud/node_controller.go]: @zetaab what do you mean we will make it every 5 seconds? Nothing I see Is on such a short interval.

Comment (zetaab) [reply pkg/controller/cloud/node_controller.go]: @jhorwit2 monitornode function is executed every 5 seconds

Comment (jhorwit2) [reply pkg/controller/cloud/node_controller.go]: That seems unnecessarily quick. The KCM marks nodes not ready after 40 seconds by default, so i'm not sure why this would need to be faster than that.

Comment (jhorwit2) [reply pkg/controller/cloud/node_controller.go]: @zetaab To your point earlier, this would cause up to 200 api calls every 5 seconds to the cloud provider since you'd not only check for if it's shutdown but also check if it's been deleted for every node that's healthy.

Comment (zetaab) [reply pkg/controller/cloud/node_controller.go]: I think we could do this whole thing little bit better. If we increase monitornode time, it will mean that removing / adding shutdown taint takes 40seconds+monitornode delay maximum. I mean that if node notready takes 40seconds, it might mean that adding taint takes 40seconds+5seconds maximum currently.

What about if we re-structure this whole feature, if node goes notready state why it cannot push message towards monitornode(using channel maybe)? Then there is (almost) no delay, and we do not need refresh that 24/7 every 5seconds?

Comment (zetaab) [reply pkg/controller/cloud/node_controller.go]: @jhorwit2 yep 200 maximum per 5seconds with 100 nodes. @gmarek well we could do that. However, I would like to modify then this whole monitornode function. Instead of checking nodes status one by one in cloudprovider, we should make bulk query towards cloudprovider (which input contains all nodes providerids). Then cloudprovider queries providerids statuses from cloudprovider API, and returns three arrays as return to this monitornode function. These arrays are: shutdown, active and removed. Where shutdown means = add taint if taint not exist yet. Active = if taint exist remove taint, removed = delete node

File: pkg/controller/cloud/node_controller.go
Comment (zetaab) [reply pkg/controller/cloud/node_controller.go]: @yastij tests added now

**Comment by zetaab** [comment]
note: gce and azure vendor libraries does not support retrieving instance state. AWS+openstack does and I have code for those already. This means that we cannot have gce+azure support for this taint thing before vendor libraries are updated.

**Comment by zetaab** [comment]
@jingxu97 @khenidak in issue of this PR you said that taints should be added in node lifecycle controller instead of cloud controller? Do you mean that I should remove taint things from cloud controller and move those to similar place here https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L668 (and then implement isShutdown(nodeName) to cloudprovider interface). Or what you think about this PR is it valid and can be approved or need changes?

**Comment by khenidak** [comment]
@zetaab 
> ..  in issue of this PR you said that taints should be added in node lifecycle controller instead of cloud controller? Do you mean that I should remove taint things from cloud controller and move those to similar place here https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L668 (and then implement isShutdown(nodeName) to cloudprovider interface)

**Yes**

> .. Or what you think about this PR is it valid and can be approved or need changes?

Yes - **Reservation** - not blocking merging this:
5 sec is too quick to check a node complete reboot cycle. Specially if the host is under updates. We are forcing everybody to modify controller manager's `--node-monitor-period` because the default value is too small. 

We have `--TaintNodesByCondition` which logically does the same thing ( i realize this covers only disk/mem/cpu). Maybe it should cover *not running* condition as well. *I will try to find a way to bring this node sig.*

> re Arm/SDK does not have running/shutdown state. 

I think it does instance view have this data:
https://resources.azure.com/subscriptions/{SUB-ID}/resourceGroups/{RG-NAME}/providers/Microsoft.Compute/virtualMachines/{VM-NAME}/instanceView

and the sdk have it [here] (https://github.com/Azure/azure-sdk-for-go/blob/master/arm/compute/virtualmachines.go#L638)

**Comment by khenidak** [comment]
Folks - Can we go ahead and merge this to give cloud provider a chance to implement the needed interfaces?

**Comment by khenidak** [comment]
@jdumars  for tracking. 1.10 will have new interface for shutdown taints
/sig @kubernetes/sig-azure-misc

**Comment by zetaab** [comment]
Why I modified nodelifecycle controller as well? When we use kubernetes we currently can decide between two node controllers: cloud controller manager or kube controller can handle nodes. The problem in this PR was that modifications was made only to cloud controller manager which is in alpha(?) and its not used yet in many places. However, I decided to add this taint thing to kube controller as well, and it means that we will get taint support faster to volumes. Is this bad thing or what you guys think? Now it is added to all node controllers, it should work in similar way no matter do you use cloud controller manager or kube controller.

**Comment by zetaab** [comment]
could you guys review this PR? If this is going to be in 1.10 we must give some time to implement cloudproviders part as well. I can do AWS and openstack but it cannot be done before this is merged.

File: pkg/controller/cloud/node_controller.go
Comment (sjenning) [inline pkg/controller/cloud/node_controller.go]: nit: should `exists` be something like `shutdown`? 

If you do decided to change this, you don't have to change 245/271 below.

**Comment by jsafrane** [comment]
/approve 
volume changes

**Comment by dims** [comment]
/approve

cloud provider changes

**Comment by jhorwit2** [comment]
/assign @deads2k

for controller approval

**Comment by bsalamat** [comment]
/approve
for the change in pkg/scheduler

**Comment by jingxu97** [comment]
@zetaab Sorry for sending this late. But I have one concern about it. My understanding is both node_controller in cloudprovider and node_lifecycle_controller checks nodeExistsInCloudProvider and might delete the node. Either controller might delete the node first. For Cinder/Aws, if node shuts down, cloudprovider will return false for nodeExistsInCloudProvider call. So if node_lifecycle_controlle checks first, it will delete the node directly without taint the node. 
Now the problem is that volume manager will depend on the taint to determine whether mounts are gone or not. So if node is deleted without the taint, volume manager will still think the mounts exist.
cc @yujuhong

**Comment by aleksandra-malinowska** [comment]
Currently [all Cluster Autoscaling tests](https://k8s-testgrid.appspot.com/sig-autoscaling#gci-gce-autoscaling) are failing due to nodes becoming unready, but never being removed. I've found the following in [kube-controller-manager.log](https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-autoscaling/6055/artifacts/bootstrap-e2e-master/cluster-autoscaler.log) from test run for one such node:

`E0216 01:14:27.084923       1 node_lifecycle_controller.go:686] Error determining if node bootstrap-e2e-minion-group-01b1 shutdown in cloud: failed to get instance ID from cloud provider: instance not found`

This line keeps appearing for the rest of the run (~6h). Looking at the code, it seems the node is stuck in Unready state because of this check: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L684 and the part of the code where it can removed is never reached.

**Comment by zetaab** [comment]
@aleksandra-malinowska, right its bug. Here https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L685 here we ignore only cloudprovider.NotImplemented error. However, there can be other errors as well which should be ignored https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/util/node/controller_utils.go#L199

**Comment by zetaab** [comment]
@jingxu97 I am checking that shutdown thing in node_lifecycle_controller first https://github.com/zetaab/kubernetes/blob/nodetaint/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L682-L714 nodeShutdownInCloudProvider. So it should not be a problem? In both controllers I am doing that. However, this PR is coming back there was one bug that is mentioned above.

**Comment by justinsb** [comment]
When we reapply this, let's review it alongside at least some cloudprovider implementation changes:

* I think this interface requires two API calls where previously we needed one, so there may be quota issues.  I'd prefer to see a call that returns another interface that represents the instance state, personally - I think that would help us avoid an explosion of these methods.
* I'd also personally rather see us passing in the api.Node object if we have it, rather than passing in `node.Spec.ProviderID`.  That can help cloudproviders avoid extra API calls - though ProviderID is usually the one we want, we've switched to Node in the past and it's been helpful.
* I'd rather call the method (whether it on a new `NodeState` interface or on the top-level interface) something more explicit, like `ShouldTaintWithShutDownTaint`.
* The really important question... ShutDown or Shutdown :-)

**Comment by zetaab** [comment]
@justinsb good points. Actually I was waiting when someone says that we could do this in one interface call instead of two different :) I will post my proposal later today about modifying cloudprovider interface

**Comment by zetaab** [comment]
@justinsb @jhorwit2 @jingxu97 what you think about interface like this https://github.com/zetaab/kubernetes/blob/proposal/pkg/cloudprovider/cloud.go#L151-L154 by using that, we could get clearly states in one api query what we need. Exists = no need for action, InstanceShouldRemove = cloud/nodelifecycle controller takes action and will delete node, InstanceShutdown = taint is added. In future we can also add states to this function.

Example what it looks like in openstack https://github.com/zetaab/kubernetes/blob/proposal/pkg/cloudprovider/providers/openstack/openstack_instances.go#L149-L170 

and cloud controller (not maybe perfect yet) https://github.com/zetaab/kubernetes/blob/proposal/pkg/controller/cloud/node_controller.go#L250-L305 

please note that this is work in progress version (I have not even compiled it). Also I will add similar structure to nodelifecycle controller that it works in similar way.

File: pkg/controller/cloud/node_controller.go
Comment (jingxu97) [inline pkg/controller/cloud/node_controller.go]: It looks like we cannot skip here.

File: pkg/controller/cloud/node_controller.go
Comment (jingxu97) [inline pkg/controller/cloud/node_controller.go]: If shutdown state returns without err, taint the node. Otherwise, continue checking

**Comment by zetaab** [comment]
@jingxu97  I cannot continue pushing to this PR before it is reverted. However, I did changes like this https://github.com/zetaab/kubernetes/commit/44e63d4e609f274410642934276bf01be7ce7d3d both controllers are now like you mentioned? Then in nodelifecycle controller, I am not calling api anymore to get ProviderID instead I am using node.Spec.ProviderID. This will remove one api call. 

Still we could do this better, but is there sense to do big change? We could combine these two api calls to one https://github.com/zetaab/kubernetes/blob/nodetaint/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L684 and https://github.com/zetaab/kubernetes/blob/nodetaint/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L696 but like I said its big change, and I am not sure do I want to make that. I have already one revert in my CV :)

**Pull request merged**
